reactor-tools/src/main/java/reactor/tools/agent/ReactorDebugAgent.java

```diff
@@ -0,0 +1,222 @@
+				if (name.startsWith("sun.")) continue;
+				if (name.startsWith("jdk.")) continue;
+				if (name.startsWith("reactor.core.")) continue;
+
+				instrumentation.retransformClasses(aClass);
+			}
+			catch (NoClassDefFoundError ignored) {
+				// Ignored
+			}
+			catch (Throwable e) {
+				System.err.println("Failed to retransform " + aClass.getName() + ":");
+				e.printStackTrace();
+				// Some classes fail to re-transform (e.g. worker.org.gradle.internal.UncheckedException)
+				// See https://bugs.openjdk.java.net/browse/JDK-8014229
+			}
+		}
+	}
+
+	static class AssemblyInfoAddingMethodVisitor extends MethodVisitor {
+
+		private final String currentMethod;
+
+		private final String currentClassName;
+
+		private final String currentSource;
+
+		private int currentLine = -1;
+
+		AssemblyInfoAddingMethodVisitor(MethodVisitor visitor, String currentMethod, String currentClassName, String currentSource) {
+			super(Opcodes.ASM7, visitor);
+			this.currentMethod = currentMethod;
+			this.currentClassName = currentClassName;
+			this.currentSource = currentSource;
+		}
+
+		@Override
+		public void visitLineNumber(int line, Label start) {
+			super.visitLineNumber(line, start);
+			currentLine = line;
+		}
+
+		@Override
+		public void visitMethodInsn(int opcode, String owner, String name, String descriptor, boolean isInterface) {
+			super.visitMethodInsn(opcode, owner, name, descriptor, isInterface);
+			switch (owner) {
+				case "reactor/core/publisher/Flux":
+				case "reactor/core/publisher/Mono":
+				case "reactor/core/publisher/ParallelFlux":
+					String returnType = descriptor.substring(descriptor.lastIndexOf(")") + 1);
+					if (!returnType.startsWith("Lreactor/core/publisher/")) {
+						return;
+					}
+
+					String callSite = String.format(
+							"%s.%s\n%s.%s(%s:%d)",
+							owner.replace("/", "."), name,
+							currentClassName.replace("/", "."), currentMethod, currentSource, currentLine
+					);
+					super.visitLdcInsn(callSite);
+					super.visitMethodInsn(
+							Opcodes.INVOKESTATIC,
+							"reactor/core/publisher/Hooks",
+							"addAssemblyInfo",
+							"(Lreactor/core/CorePublisher;Ljava/lang/String;)Lreactor/core/CorePublisher;",
+							false
+					);
+					break;
+			}
+		}
+	}
+
+	static class PublicMethodsClassFileTransformer implements ClassFileTransformer {
+
+		@Override
+		public byte[] transform(
+				ClassLoader loader,
+				String className,
+				Class<?> clazz,
+				ProtectionDomain protectionDomain,
+				byte[] bytes
+		) {
+			if ("reactor/core/publisher/Hooks".equals(className)) {
+				ClassReader cr = new ClassReader(bytes);
+				ClassWriter cw = new ClassWriter(cr, 0);
+
+				ClassVisitor classVisitor = new ClassVisitor(Opcodes.ASM7, cw) {
+					@Override
+					public MethodVisitor visitMethod(int access,
+							String name,
+							String descriptor,
+							String signature,
+							String[] exceptions) {
+						if ("addAssemblyInfo".equals(name)) {
+							access |= Opcodes.ACC_PUBLIC;
```

Comment:
does it make the method public? if so, maybe the same strategy can be applied to `Traces` and we could put that `addAssemblyInfo` in `Traces`?

Reply:
It does (see the implementation notes in the PRs description), also see my reply here:
https://github.com/reactor/reactor-core/pull/1553#discussion_r265491071